libs/lbox-clients/src/lbox/request_client.py

```diff
@@ -0,0 +1,368 @@
+from lbox import exceptions  # type: ignore
+
+logger = logging.getLogger(__name__)
+
+_LABELBOX_API_KEY = "LABELBOX_API_KEY"
+
+
+def python_version_info():
+    version_info = sys.version_info
+
+    return f"{version_info.major}.{version_info.minor}.{version_info.micro}-{version_info.releaselevel}"
+
+
+class RequestClient:
+    """A Labelbox request client.
+
+    Contains info necessary for connecting to a Labelbox server (URL,
+    authentication key).
+    """
+
+    def __init__(
+        self,
+        sdk_version,
+        api_key=None,
+        endpoint="https://api.labelbox.com/graphql",
+        enable_experimental=False,
+        app_url="https://app.labelbox.com",
+        rest_endpoint="https://api.labelbox.com/api/v1",
+    ):
+        """Creates and initializes a RequestClient.
+        This class executes graphql and rest requests to the Labelbox server.
+
+        Args:
+            api_key (str): API key. If None, the key is obtained from the "LABELBOX_API_KEY" environment variable.
+            endpoint (str): URL of the Labelbox server to connect to.
+            enable_experimental (bool): Indicates whether or not to use experimental features
+            app_url (str) : host url for all links to the web app
+        Raises:
+            exceptions.AuthenticationError: If no `api_key`
+                is provided as an argument or via the environment
+                variable.
+        """
+        if api_key is None:
+            if _LABELBOX_API_KEY not in os.environ:
+                raise exceptions.AuthenticationError("Labelbox API key not provided")
+            api_key = os.environ[_LABELBOX_API_KEY]
+        self.api_key = api_key
+
+        self.enable_experimental = enable_experimental
+        if enable_experimental:
+            logger.info("Experimental features have been enabled")
+
+        logger.info("Initializing Labelbox client at '%s'", endpoint)
+        self.app_url = app_url
+        self.endpoint = endpoint
+        self.rest_endpoint = rest_endpoint
+        self.sdk_version = sdk_version
+        self._connection: requests.Session = self._init_connection()
+
+    def _init_connection(self) -> requests.Session:
+        connection = requests.Session()  # using default connection pool size of 10
+        connection.headers.update(self._default_headers())
+
+        return connection
+
+    @property
+    def headers(self) -> MappingProxyType:
+        return self._connection.headers
+
+    def _default_headers(self):
+        return {
+            "Authorization": "Bearer %s" % self.api_key,
+            "Accept": "application/json",
+            "Content-Type": "application/json",
+            "X-User-Agent": f"python-sdk {self.sdk_version}",
+            "X-Python-Version": f"{python_version_info()}",
+        }
+
+    @retry.Retry(
+        predicate=retry.if_exception_type(
+            exceptions.InternalServerError,
+            exceptions.TimeoutError,
+        )
+    )
+    def execute(
```

Comment:
just wondering. Is there a plan to some day refactor this beast of a function? I get dizzy each time i have to scroll through it 😄 I believe, there are few oop patterns (strategy, visitor) that could help dismantled this function and make it more simple and facilitate maintenance in the long run

Reply:
tes @mihhail-m we will refactor it ... some time :)